plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/rest/TrinoIcebergRESTCatalogFactory.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg.catalog.rest;
+
+import io.airlift.log.Logger;
+import io.trino.hdfs.ConfigurationUtils;
+import io.trino.plugin.base.CatalogName;
+import io.trino.plugin.hive.NodeVersion;
+import io.trino.plugin.iceberg.IcebergConfig;
+import io.trino.plugin.iceberg.catalog.TrinoCatalog;
+import io.trino.plugin.iceberg.catalog.TrinoCatalogFactory;
+import io.trino.spi.security.ConnectorIdentity;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.rest.RESTSessionCatalog;
+import org.apache.iceberg.rest.auth.OAuth2Properties;
+
+import javax.inject.Inject;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class TrinoIcebergRESTCatalogFactory
+        implements TrinoCatalogFactory
+{
+    private static final Logger log = Logger.get(TrinoIcebergRESTCatalogFactory.class);
+    private final CatalogName catalogName;
+    private final String trinoVersion;
+    private final String serverUri;
+    private final Optional<String> credential;
+    private final Optional<String> token;
+
+    private volatile RESTSessionCatalog icebergCatalog;
+
+    @Inject
+    public TrinoIcebergRESTCatalogFactory(
+            CatalogName catalogName,
+            IcebergConfig icebergConfig,
+            IcebergRESTCatalogConfig restConfig,
+            NodeVersion nodeVersion)
+    {
+        this.catalogName = catalogName;
+        this.trinoVersion = requireNonNull(nodeVersion, "nodeVersion is null").toString();
+        this.serverUri = restConfig.getBaseUri();
+        this.credential = restConfig.getCredential();
+        this.token = restConfig.getToken();
+    }
+
+    @Override
+    public TrinoCatalog create(ConnectorIdentity identity)
+    {
+        if (icebergCatalog == null) {
+            synchronized (this) {
+                if (icebergCatalog == null) {
+                    Map<String, String> properties = new HashMap<>();
+                    properties.put(CatalogProperties.URI, serverUri);
+                    properties.put(CatalogProperties.FILE_IO_IMPL, "org.apache.iceberg.io.ResolvingFileIO");
+                    properties.put("trino-version", trinoVersion);
```

Comment:
Q How does Iceberg use `trino-version` property? Setting in the request header? It worth to leave a code comment because it's not intuitive.

Reply:
The trino version passed along to the REST Catalog so that we know which client version is communicating with the backend.  There isn't a specific use case currently, but if there are any issues identified with specific versions, it can be addressed with checks on the server side.